angularSTEP/src/app/upload-recipe/upload-recipe.component.ts

```diff
@@ -11,9 +14,48 @@ import {Router} from '@angular/router';
   styleUrls: ['./upload-recipe.component.scss']
 })
 export class UploadRecipeComponent{
+  error = '';
+  hide = true;
+  imgFile!: Blob;
+  pictureUrl = '';
+  progress = 0;
+  previewImgUrls: SafeUrl[] = [];
+  imageFiles: Blob[] = [];
+  @Output() uploadDone = new EventEmitter<boolean>();
+
   constructor(private fb: FormBuilder,
               private db: AngularFirestore,
-              private router: Router,) { }
+              private router: Router,
+              private storage: AngularFireStorage,
+              private afs: AngularFirestore,
+              private _ngZone: NgZone,
+              private sanitizer: DomSanitizer,
+              private fAuth: AngularFireAuth) {}
+
+  fileFormGroup = new FormGroup({
+    imageArray: new FormArray([], {validators: Validators.required} )
+  });
+
+  get imageArray() {
+    return this.fileFormGroup.get('imageArray') as FormArray;
+  }
+
+  deleteImage(i: number) {
+    this.previewImgUrls.splice(i, 1);
+    this.imageArray.removeAt(i);
+  }
+
+  addImageToArray(event: any) {
+    if (event.target.files.length > 0) {
+      const file: Blob = event.target.files[0];
+      this.imgFile = file;
+      //uploads preview of image
+      this.previewImgUrls.push(this.sanitizer.bypassSecurityTrustUrl(
```

Comment:
"bypassSecutiryTrustUrl" is a scary thing to read. I think I get why this is OK: you are only explicitly allowing a URL that points to a known thing in your DB, not loosening security on the data blob itself. If that's right, it would be nice to have a comment here saying that.

Reply:
Oh yes, I should definitely comment that on what it means. What happens is to display the preview of the image, it creates a temporary URL for the known blob. This image/blob is one the user has just uploaded. It does not change the actual security of the URL, but allows the website to just display it temporarily. You could do it manually to any image by deleting the first couple characters of the URL, but this just does that for the website.